sources/packages/backend/libs/services/src/msfaa-number/msfaa-number-shared.service.ts

```diff
@@ -0,0 +1,266 @@
+    );
+    return this.internalCreateMSFAANumber(
+      application.student.id,
+      application.id,
+      application.currentAssessment.offering.offeringIntensity,
+      auditUserId,
+    );
+  }
+
+  /**
+   * Reissues a new MSFAA number to be associated with the student, cancelling any
+   * pending MSFAA for the particular offering intensity and also associating the
+   * new MSFAA number to any pending disbursement for the same offering intensity.
+   * Reissuing an MSFAA is only valid for an application with pending disbursement
+   * where the MSFAA is cancelled.
+   * @param referenceApplicationId reference application id.
+   * @returns the newly created MSFAA.
+   */
+  async reissueMSFAA(
+    referenceApplicationId: number,
+    auditUserId: number,
+  ): Promise<MSFAANumber> {
+    const application = await this.getApplicationForMSFAACreation(
+      referenceApplicationId,
+    );
+    const pendingDisbursement =
+      application.currentAssessment.disbursementSchedules.find(
+        (schedule) =>
+          schedule.disbursementScheduleStatus ===
+          DisbursementScheduleStatus.Pending,
+      );
+    if (!pendingDisbursement) {
+      throw new CustomNamedError(
+        "Not possible to reissue an MSFAA when there is no pending disbursements for the application.",
+        APPLICATION_INVALID_DATA_TO_CREATE_MSFAA_ERROR,
+      );
+    }
+    if (!pendingDisbursement.msfaaNumber.cancelledDate) {
+      throw new CustomNamedError(
+        "Not possible to reissue an MSFAA when the current associated MSFAA is not cancelled.",
+        APPLICATION_INVALID_DATA_TO_CREATE_MSFAA_ERROR,
+      );
+    }
+    return this.internalCreateMSFAANumber(
+      application.student.id,
+      application.id,
+      application.currentAssessment.offering.offeringIntensity,
+      auditUserId,
+    );
+  }
+
+  /**
+   * Creates a new MSFAA record with a new number for the specified student.
+   * @param studentId student to have a new MSFAA record created.
+   * @param referenceApplicationId reference application id.
+   * @param offeringIntensity offering intensity.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * individually based on, for instance, the Part time/Full time.
+   * @returns created MSFAA record.
+   */
+  private async internalCreateMSFAANumber(
+    studentId: number,
+    referenceApplicationId: number,
+    offeringIntensity: OfferingIntensity,
+    auditUserId: number,
+  ): Promise<MSFAANumber> {
+    return this.dataSource.transaction(async (entityManager) => {
+      const auditUser = await this.systemUsersService.systemUser();
+      const now = new Date();
+      const nowISODate = getISODateOnlyString(now);
+      // Cancel any pending MSFAA record that was never reported as signed or cancelled.
+      // This will ensure that the newly created MSFAA will be the only one valid.
+      await entityManager.getRepository(MSFAANumber).update(
```

Comment:
I see that is is a very rare edge case scenario where 

Student submit an application where MSFAA was valid during submission and before the disbursement was sent it expired and a new application got created with new MSFAA in the mean time. 

Hence when reissuing msfaa we are checking for this MSFAA which was never reported or cancelled.

But I have a question here, while re-issuing should we not use this non-reported msfaa number? Is there anything like a policy which stops us doing it?

Reply:
I was trying to recall the exact reason to have them canceled actually. The goal was to ensure that when a new MSFAA is created there will be only one "not signed".
For instance, the student created an application, generated an MSFAA, and never completed it. After 3 years the same student comes back and a new MSFAA will be generated, in this situation that old record will be "cancelled by the system".
I believe that the idea was just to keep the DB looking better with only one possible not-signed/not-canceled MSFAA.